File: go/problem_55.go
package main

import (
	"crypto/rand"
	"encoding/hex"
)

type URLShortener struct {
	short2long map[string]string
	long2short map[string]string
}

func newURLShortener() *URLShortener {
	return &URLShortener{
		short2long: map[string]string{},
		long2short: map[string]string{},
	}
}

func (u *URLShortener) shorten(url string) string {
	if s, ok := u.long2short[url]; ok {
		return s
	}
	b := make([]byte, 3)
	rand.Read(b)
	s := hex.EncodeToString(b)
	u.short2long[s] = url
	u.long2short[url] = s
	return s
}

func (u *URLShortener) restore(short string) (string, bool) {
	v, ok := u.short2long[short]
	return v, ok
}

func problem55() {
	println("Problem 55:")
	us := newURLShortener()
	s1 := us.shorten("https://example.com")
	if len(s1) == 0 {
		panic("empty short")
	}
	orig, ok := us.restore(s1)
	if !ok || orig != "https://example.com" {
		panic("restore failed")
	}
	s2 := us.shorten("https://example.com")
	if s1 != s2 {
		panic("same URL should map to same short")
	}
	_, ok = us.restore("notexist")
	if ok {
		panic("should not exist")
	}
	println("All Tests Passed")
}

File: c/problem_55.c
#include "problem_55.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>

void url_init(URLShortener55 *us) { us->size = 0; }

const char *url_shorten(URLShortener55 *us, const char *url) {
    for (int i=0; i<us->size; i++)
        if (strcmp(us->orig[i], url)==0) return us->shortened[i];
    int idx = us->size++;
    strncpy(us->orig[idx], url, 255);
    snprintf(us->shortened[idx], SHORT_LEN, "%06x", (unsigned)(rand() & 0xFFFFFF));
    return us->shortened[idx];
}

const char *url_restore(URLShortener55 *us, const char *s) {
    for (int i=0; i<us->size; i++)
        if (strcmp(us->shortened[i], s)==0) return us->orig[i];
    return NULL;
}

void problem55(void) {
    printf("Problem 55:\n");
    URLShortener55 us; url_init(&us);
    const char *s1 = url_shorten(&us, "https://example.com");
    assert(s1 && strlen(s1) > 0);
    assert(strcmp(url_restore(&us, s1), "https://example.com")==0);
    assert(strcmp(url_shorten(&us, "https://example.com"), s1)==0);
    assert(url_restore(&us, "notexist")==NULL);
    printf("All Tests Passed\n");
}
